### DISCUSSION TITLE
Support LibMan package manager

### DISCUSSION THREAD

**Discussion by odin568** [renovatebot/renovate#31250] [Suggest an Idea]
### Tell us more.

Coming from https://github.com/renovatebot/renovate/issues/2642
Yes, I use it :) 
We us it in order to have a central place to keep track of versions of client-side libraries like jquery, etc.
We have asp.net core MVC application. 
For various reasons, we don't use CDNs but ship everything with our application.

Before we had jquery checked in for example in the wwwroot folder. This is odd - nobody knows which version it is, etc.

Here comes Libman, which defines versions and a dotnet build automatically retrieves the library from CDN and puts it into wwwroot. This happens on the fly - the folder is excluded via gitignore. So binaries are not checked in.

I was coming from dependabot but now switching to renovate and would love to see support for this.

An example libman.json from us:

```json
{
  "version": "1.0",
  "defaultProvider": "cdnjs", // Check manually for updates from time to time via https://cdnjs.com/
  "libraries": [
    {
      "provider": "cdnjs",
      "library": "bootstrap@5.3.3",
      "destination": "wwwroot/lib/bootstrap/"
    },
    {
      "provider": "cdnjs",
      "library": "bootstrap-icons@1.11.3",
      "destination": "wwwroot/lib/bootstrap-icons/"
    },
    {
      "provider": "cdnjs",
      "library": "flag-icons@7.2.3",
      "destination": "wwwroot/lib/flag-icons/"
    },
    {
      "provider": "cdnjs",
      "library": "highlight.js@11.10.0",
      "destination": "wwwroot/lib/highlight.js/"
    },
    {
      "provider": "cdnjs",
      "library": "jquery@3.7.1",
      "destination": "wwwroot/lib/jquery/"
    },
    {
      "provider": "cdnjs",
      "library": "jquery-validate@1.21.0",
      "destination": "wwwroot/lib/jquery-validate/"
    },
    {
      "provider": "cdnjs",
      "library": "jquery-validation-unobtrusive@4.0.0",
      "destination": "wwwroot/lib/jquery-validation-unobtrusive/"
    },
    {
      "provider": "cdnjs",
      "library": "jszip@3.10.1",
      "destination": "wwwroot/lib/jszip/"
    },
    {
      "provider": "unpkg",
      "library": "@progress/kendo-font-icons@3.0.0", // Current version check https://unpkg.com/@progress/kendo-font-icons/package.json
      "destination": "wwwroot/lib/progress/kendo-font-icons/",
      "files": [
        "**/*",
        "!*.md",
        "!package.json"
      ]
    }
  ]
}
```

**Comment by rouke-broersma** [comment]
I built a custom regex matcher for libman today, enjoy! It does not use the selected package source for each dependency, but as all of the sources iirc are basically file caches for npm I assume only using npm as the source of the update check works fine.

Example:

```
  "customManagers": [
    {
      "fileMatch": ["(^|/)libman\\.json$"],
      "customType": "regex",
      "datasourceTemplate": "npm",
      "matchStrings": [
        "\\\"library\\\":\\s?\\\"(?<packageName>.*)@(?<currentValue>.*)\\\""
      ]
    }
  ],
```

**Comment by odin568** [comment]
oh that's brilliant!! Thank you very much! 
I will play around with it but looks good, also with npm as source. IMHO would have deserved it as a standard 😄

**Reply by rouke-broersma** [reply]
If I have some time and can figure out how to contribute this to renovate I'll see what I can do

**Comment by odin568** [comment]
Ah, one thing I need to dig in (but now it's weekend first :) )
It detects for jquery-validate an update to 2.0.0 which routes then to deprecate holder https://github.com/npm/deprecate-holder
Not sure what this is about. AFAIK it is not deprecated but need to look. On cdnjs, there is no 2.0.0

**Reply by rouke-broersma** [reply]
https://www.npmjs.com/package/jquery-validate

This package does in fact seem to be deprecated, however it seems like on cdnjs they have decided to instead of deprecating replace the package with jquery-validation instead..? See: https://www.npmjs.com/package/jquery-validation

Just naming the package something else on a cdn is quite bad in my opinion, I would consider using the one from jsdelivr which has the correct name: https://jqueryvalidation.org/#latest-files-on-jsdelivr-cdn-(hotlinking-welcome):

**Reply by odin568** [reply]
Okay, thanks for your help. That's of course tricky if a cdn renames dependencies. If a package name changes, I would prefer that it breaks rather than doing some magic, but that might be a design question.

I switched now to jsdelivr and everything works fine. Thanks a lot for your help!

**Comment by Piedone** [comment]
Here's our take on a `jsonata` custom manager to handle LibMan library definitions: https://github.com/Lombiq/renovate-config/blob/d5cdf2c3d2d3b4d308bd20f8c021039e65ab89a8/default.json5#L53
